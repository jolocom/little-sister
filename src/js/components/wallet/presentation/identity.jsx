import React from 'react'
import Radium from 'radium'

import { TextField, Divider, List, ListItem, Avatar } from 'material-ui'

import Loading from 'components/common/loading'
import CommunicationCall from 'material-ui/svg-icons/communication/call'
import CommunicationEmail from 'material-ui/svg-icons/communication/email'
import CameraIcon from 'material-ui/svg-icons/image/photo-camera'

import {theme} from 'styles'
import {Content, Block} from '../../structure'
import {
  PlusMenu,
  TabContainer,
  HalfScreenContainer,
  ContactList,
  IdCardsList,
  PassportsList,
  InfoDetails
} from './ui'

const STYLES = {
  listItem: {
    whiteSpace: 'nowrap',
    padding: '0 16px 0 72px'
  },
  inputName: {
    color: theme.palette.textColor,
    fontSize: '1em'
  },
  labelName: {
    color: theme.textStyles.labelInputFields.color
  },
  divider: {
    marginLeft: '16px'
  },
  simpleDialog: {
    contentStyle: {
    },
    actionsContainerStyle: {
      textAlign: 'right'
    }
  },
  avatar: {
    marginTop: '10px'
  },
  container: {
    marginLeft: '10px',
    marginRight: '10px'
  }
}

@Radium
export default class WalletIdentity extends React.Component {
  static propTypes = {
    children: React.PropTypes.node,
    username: React.PropTypes.object.isRequired,
    passports: React.PropTypes.array,
    showUserInfo: React.PropTypes.func.isRequired,
    idCards: React.PropTypes.array,
    isLoaded: React.PropTypes.bool.isRequired,
    isError: React.PropTypes.bool.isRequired,
    webId: React.PropTypes.string.isRequired,
    phones: React.PropTypes.array.isRequired,
    emails: React.PropTypes.array.isRequired,
    goToContactManagement: React.PropTypes.func.isRequired,
    goToPassportManagement: React.PropTypes.func.isRequired,
    goToDrivingLicenceManagement: React.PropTypes.func.isRequired,
    onConfirm: React.PropTypes.func.isRequired,
    setFocusedPin: React.PropTypes.func.isRequired,
    changePinValue: React.PropTypes.func.isRequired,
    requestVerificationCode: React.PropTypes.func.isRequired,
    resendVerificationCode: React.PropTypes.func.isRequired,
    enterVerificationCode: React.PropTypes.func.isRequired,
    onVerify: React.PropTypes.func.isRequired,
    saveToBlockchain: React.PropTypes.func.isRequired
  }

  render() {
    const {
      username,
      passports,
      idCards,
      isLoaded,
      webId,
      showUserInfo,
      phones,
      emails,
      goToContactManagement,
      goToPassportManagement,
      goToDrivingLicenceManagement,
      onConfirm,
      changePinValue,
      setFocusedPin,
      requestVerificationCode,
      resendVerificationCode,
      enterVerificationCode,
      saveToBlockchain
    } = this.props

    if (!isLoaded) {
      return <Loading />
    }

    const avatar = (<Avatar
      icon={<CameraIcon viewBox="-3 -3 30 30" />}
      color={theme.jolocom.gray1}
      backgroundColor={theme.jolocom.gray3}
      style={STYLES.avatar} />)

    return (<TabContainer>
      <HalfScreenContainer>
        <Content style={STYLES.container}>
          <Block>
            <List>
              <ListItem
                key={1}
                disabled
                rightIcon={<InfoDetails
                  showDetails={details => showUserInfo(
                    null,
                    details,
                    'OK',
                    STYLES.simpleDialog
                  )}
                  webId={webId}
                  username={username.value} />
                }
                leftAvatar={avatar}
                style={STYLES.listItem}>
                <TextField
                  fullWidth
                  floatingLabelText="Name"
                  inputStyle={STYLES.inputName}
                  floatingLabelStyle={STYLES.labelName}
                  underlineShow={false}
                  floatingLabelFixed
                  value={username.value} />
              </ListItem>
              <Divider style={STYLES.divider} />
            </List>
          </Block>
          <Block>
            <PlusMenu
              name="Contact"
              choice={[...emails, ...phones].length > 0}
              goToManagement={goToContactManagement} />
          </Block>
          <Block>
            <ContactList
              fields={phones}
              changePinValue={changePinValue}
              onConfirm={onConfirm}
              icon={CommunicationCall}
              setFocusedPin={setFocusedPin}
              requestVerificationCode={requestVerificationCode}
              resendVerificationCode={resendVerificationCode}
              enterVerificationCode={enterVerificationCode}
              labelText="Phone Number"
              attrType="phone" />
            <ContactList
              fields={emails}
              onConfirm={onConfirm}
              requestVerificationCode={requestVerificationCode}
              resendVerificationCode={resendVerificationCode}
              enterVerificationCode={enterVerificationCode}
              icon={CommunicationEmail}
              labelText="Email"
              attrType="email" />
          </Block>
          <Block>
            <PlusMenu
              name="Passport"
              choice={passports.length > 0}
              goToManagement={goToPassportManagement} />
          </Block>
          <Block>
            <PassportsList passports={passports} />
          </Block>
          <Block>
            <PlusMenu
              name="ID Card"
              choice={idCards.length > 0}
              goToManagement={goToPassportManagement} />
          </Block>
          <Block>
            <IdCardsList
              idCards={idCards}
              saveToBlockchain={saveToBlockchain} />
          </Block>
          <Block>
            <PlusMenu
              name="Driving License"
              choice={false}
              goToManagement={goToDrivingLicenceManagement} />
          </Block>
        </Content>
      </HalfScreenContainer>
    </TabContainer>)
  }
}
