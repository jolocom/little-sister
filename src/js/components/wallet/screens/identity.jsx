import React from 'react'
import {connect} from 'redux/utils'
import Presentation from '../presentation/identity'
import WalletError from '../presentation/error'

@connect({
  props: ['wallet'],
  actions: [
    'wallet/identity:getIdentityInformation',
    'wallet/identity:changeSmsCodeValue',
    'wallet/identity:changePinValue',
    'wallet/identity:setFocusedPin',
    'wallet/identity:goToPassportManagement',
    'wallet/identity:goToDrivingLicenceManagement',
    'wallet/identity:goToContactManagement',
    'confirmation-dialog:openConfirmDialog',
    'confirmation-dialog:closeConfirmDialog',
    'simple-dialog:showSimpleDialog',
    'simple-dialog:configSimpleDialog',
    'verification:confirmEmail',
    'verification:confirmPhone',
    'verification:startEmailVerification',
    'verification:startPhoneVerification',
    'wallet/identity:saveToBlockchain'
  ]
})

export default class WalletIdentityScreen extends React.Component {
  static propTypes = {
    children: React.PropTypes.node,
    wallet: React.PropTypes.object,
    goToPassportManagement: React.PropTypes.func.isRequired,
    setFocusedPin: React.PropTypes.func.isRequired,
    goToDrivingLicenceManagement: React.PropTypes.func.isRequired,
    goToContactManagement: React.PropTypes.func.isRequired,
    closeConfirmDialog: React.PropTypes.func.isRequired,
    openConfirmDialog: React.PropTypes.func.isRequired,
    getIdentityInformation: React.PropTypes.func.isRequired,
    configSimpleDialog: React.PropTypes.func.isRequired,
    showSimpleDialog: React.PropTypes.func.isRequired,
    startPhoneVerification: React.PropTypes.func.isRequired,
    startEmailVerification: React.PropTypes.func.isRequired,
    confirmEmail: React.PropTypes.func.isRequired,
    confirmPhone: React.PropTypes.func.isRequired,
    resendVerificationLink: React.PropTypes.func,
    resendVerificationSms: React.PropTypes.func,
    changePinValue: React.PropTypes.func.isRequired,
    changeSmsCodeValue: React.PropTypes.func.isRequired,
    saveToBlockchain: React.PropTypes.func.isRequired
  }

  componentWillMount() {
    this.props.getIdentityInformation()
  }

  render() {
    const {username, contact, webId, passports, idCards, loaded, error
    } = this.props.wallet.identity
    if (error) {
      return (<WalletError
        message="...oops something went wrong! We were not able to load your data." // eslint-disable-line max-len
        buttonLabel="RETRY"
        onClick={() => this.render()} />)
    }
    const {emails, phones} = contact

    return (<Presentation
      username={username}
      emails={emails}
      phones={phones}
      webId={webId}
      passports={passports}
      idCards={idCards}
      isLoaded={loaded}
      isError={error}
      setFocusedPin={this.props.setFocusedPin}
      changePinValue={this.props.changePinValue}
      goToContactManagement={this.props.goToContactManagement}
      goToPassportManagement={this.props.goToPassportManagement}
      goToDrivingLicenceManagement={this.props.goToDrivingLicenceManagement}
      requestIdCardVerification={
        ({message, rightButtonLabel, leftButtonLabel, index}) =>
          this.props.openConfirmDialog(
            message,
            rightButtonLabel,
            () => { this.props.saveToBlockchain(0) },
            leftButtonLabel
          )
        }
      requestVerificationCode={(args, params) => this.showVerificationWindow(args, () => { // eslint-disable-line max-len
        return () => this.showVerificationWindow(params,
          (callbackArgs) => this.requestVerificationCode(callbackArgs))
      })}
      resendVerificationCode={(...args) => this.showVerificationWindow(...args,
        (...params) => this.resendVerificationCode(...params)
      )}
      enterVerificationCode={(...args) => this.showVerificationWindow(...args,
        (...params) => this.enterVerificationCode(...params)
      )}
      onVerify={({title, message, buttonText, style, attrValue}) => {
        this.props.configSimpleDialog(() => {
          this.props.startEmailVerification({email: attrValue})
        }, title, message, buttonText, style)
        this.props.showSimpleDialog()
      }}
      onConfirm={(...args) => { this.onConfirm(...args) }}
      showUserInfo={(...args) => {
        this.props.configSimpleDialog(...args)
        this.props.showSimpleDialog()
      }} />)
  }

  onConfirm(args, params) {
    return this.showVerificationWindow(args, () => {
      return () => this.showVerificationWindow(params,
      (callbackArgs) => this.requestVerificationCode(callbackArgs))
    })
  }

  requestVerificationCode({attrType, attrValue, index}) {
    if (attrType === 'phone') {
      return () => this.props.startPhoneVerification({phone: attrValue, index})
    } else if (attrType === 'email') {
      return () => this.props.startEmailVerification({email: attrValue, index})
    }
  }

  resendVerificationCode({attrType, attrValue, index}) {
    if (attrType === 'phone') {
      return () => this.props.resendVerificationLink({phone: attrValue, index})
    } else if (attrType === 'email') {
      return () => this.props.resendVerificationSms({email: attrValue, index})
    }
  }

  enterVerificationCode({attrType, attrValue = 'attrValue'}) {
    if (attrType === 'phone') {
      return () => this.props.confirmPhone({phone: attrValue})
    } else if (attrType === 'email') {
      return () => this.props.confirmEmail({email: attrValue})
    }
  }

  showVerificationWindow({title, message, attrValue, attrType, index, rightButtonLabel, leftButtonLabel}, callback) { // eslint-disable-line max-len
    return this.props.openConfirmDialog(
      title,
      message,
      rightButtonLabel,
      callback({attrValue, attrType, index}),
      leftButtonLabel
    )
  }
}
