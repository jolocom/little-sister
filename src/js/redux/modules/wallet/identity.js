import Immutable from 'immutable'
import { makeActions } from '../'
import * as router from '../router'
import WebIdAgent from 'lib/agents/webid'

const actions = module.exports = makeActions('wallet/identity', {
  goToContactManagement: {
    expectedParams: [],
    creator: () => {
      return (dispatch) => {
        dispatch(router.pushRoute('/wallet/identity/contact'))
      }
    }
  },
  changeSmsCodeValue: {
    expectedParams: ['value', 'index']
  },
  changePinValue: {
    expectedParams: ['value', 'index']
  },
  setFocusedPin: {
    expectedParams: ['value', 'index']
  },
  goToPassportManagement: {
    expectedParams: [],
    creator: () => {
      return (dispatch) => {
        dispatch(router.pushRoute('/wallet/identity/passport/add'))
      }
    }
  },
  goToDrivingLicenceManagement: {
    expectedParams: [],
    creator: () => {
      return (dispatch) => {
        dispatch(router.pushRoute('/wallet/identity/drivers-licence/add'))
      }
    }
  },
  goToIdentity: {
    expectedParams: [],
    creator: () => {
      return (dispatch) => {
        dispatch(router.pushRoute('/wallet/identity/'))
      }
    }
  },
  getIdentityInformation: {
    expectedParams: [],
    async: true,
    creator: (params) => {
      return (dispatch, getState, {services, backend}) => {
        dispatch(actions.getIdentityInformation.buildAction(params, () => {
          return backend.solid.getUserInformation(new WebIdAgent().getWebId())
        }))
      }
    }
  }
})

const mapBackendToState = ({webId, contact, passports, idCards}) =>
  Immutable.fromJS({
    loaded: true,
    error: false,
    webId,
    username: {value: webId.split('.')[0].split('://')[1]},
    contact: {
      emails: contact.email,
      phones: contact.phone
    },
    passports: passports,
    idCards: idCards
  })
const mapBackendToStateError =
({webId, username, contact, passports, idCards}) =>
  Immutable.fromJS({
    loaded: true,
    error: true,
    webId: {value: ''},
    username: {value: ''},
    contact: {
      emails: [],
      phones: []
    },
    passports: [],
    idCards: []
  })
const initialState = Immutable.fromJS({
  loaded: false,
  error: false,
  webId: '',
  username: {
    verified: false,
    value: ''
  },
  contact: {
    phones: [{
      type: '',
      number: '',
      verified: false,
      smsCode: '',
      pin: '',
      pinFocused: false
    }],
    emails: [{
      type: '',
      address: '',
      pin: '',
      verified: false
    }]
  },
  passports: [
    {
      number: '',
      givenName: '',
      familyName: '',
      birthDate: '',
      gender: '',
      showAddress: '',
      streetAndNumber: '',
      city: '',
      zip: '',
      state: '',
      country: '',
      verified: false
    }
  ]
})

const changeSmsCodeValue = (state, {index, value}) => {
  if (/^[0-9]{0,6}$/.test(value)) {
    return state.mergeIn(['contact', 'phones', index], {
      smsCode: value
    })
  }
  return state
}

const changePinValue = (state, {index, value}) => {
  if (/^[0-9]{0,6}$/.test(value)) {
    return state.mergeIn(['contact', 'phones', index], {
      pin: value
    })
  }
  return state
}

module.exports.default = (state = initialState, action = {}) => {
  switch (action.type) {
    case actions.getIdentityInformation.id_success:
      return mapBackendToState(action.result)

    case actions.getIdentityInformation.id_fail:
      return mapBackendToStateError(state)

    case actions.changeSmsCodeValue.id:
      return changeSmsCodeValue(state, action)

    case actions.changePinValue.id:
      return changePinValue(state, action)

    case actions.setFocusedPin.id:
      return state.mergeIn(['contact', 'phones', action.index], {
        pinFocused: action.value
      })
    default:
      return state
  }
}
