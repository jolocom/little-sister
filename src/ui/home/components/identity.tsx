import React from 'react'
import Icon from 'react-native-vector-icons/MaterialCommunityIcons'
import { TextInput, StyleSheet, TouchableOpacity } from 'react-native'
import { Container, Block } from 'src/ui/structure'
import { JolocomTheme } from 'src/styles/jolocom-theme'
import { QRcodeScanner } from 'src/ui/home/components/qrcodeScanner'

// TODO Type for E, Event is not enough
interface Props {
  scanning: boolean,
  userName: string,
  phoneNumber: string,
  emailAddress: string,
  onUserNameChange: (input: string) => void
  onPhoneNumberChange: (input: string) => void
  onEmailAddressChange: (input: string) => void
  onScannerStart: () => void
  onScannerSuccess: (e : Event) => void
  onScannerCancel: () => void
}

interface State {
}

// TODO Magic numbers
const styles = StyleSheet.create({
  textInputField: {
    flex: 1,
    width: '80%'
  },
  block: {
    marginBottom: "15%"
  },
  icon: {
    margin: "20%"
  },
  iconContainer: {
    height: 55,
    width: 55,
    borderRadius: 35,
    backgroundColor: JolocomTheme.palette.primaryColor,
    alignItems: 'center',
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.8,
    shadowRadius: 2,
    elevation: 8,
  },
  actionButtonContainer: {
    position: 'absolute',
    right: '3%',
    bottom: '5%',
    alignItems: 'flex-end'
  },
  centerText: {
    flex: 1,
    fontSize: 18,
    padding: 32,
    color: '#777',
  },
  textBold: {
    fontWeight: '500',
    color: '#000',
  },
  buttonText: {
    fontSize: 21,
    color: 'rgb(0,122,255)',
  },
  buttonTouchable: {
    padding: 16,
  }
})

// TODO SFC?
export class IdentityComponent extends React.Component<Props, State> {
  render() {
    let renderContent
    if (this.props.scanning) {
      renderContent = (
        <QRcodeScanner
          onScannerSuccess={this.props.onScannerSuccess}
          onScannerCancel={this.props.onScannerCancel}
        />
      )
    } else {
      renderContent = (
        <Block>
          <Block style={ styles.block }>
            <TextInput
              style={ styles.textInputField }
              placeholder='Please enter your name'
              onChangeText={ this.props.onUserNameChange }
              value={ this.props.userName}
            />
            <TextInput
              style={ styles.textInputField }
              placeholder='Please enter your phone number'
              onChangeText={ this.props.onPhoneNumberChange}
              value={ this.props.phoneNumber}
            />
            <TextInput
              style={ styles.textInputField }
              placeholder='Please enter your email address'
              onChangeText={ this.props.onEmailAddressChange}
              value={ this.props.emailAddress}
            />
          </Block>
          <Block style={ styles.actionButtonContainer }>
            <TouchableOpacity
              style={ styles.iconContainer }
              onPress={ this.props.onScannerStart }>
              <Icon
                style={ styles.icon }
                size={ 30 }
                name="qrcode-scan"
                color="white"
              />
          </TouchableOpacity>
        </Block>
      </Block>
      )
    }

    return (
      <Container>
        {renderContent}
      </Container>
    )
  }
}
