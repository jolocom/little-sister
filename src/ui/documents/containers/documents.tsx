import React from 'react'
import { connect } from 'react-redux'
import { Animated, ScrollView, StyleSheet, Text, View } from 'react-native'

import { openDocumentDetails } from 'src/actions/documents'
import { DecoratedClaims } from 'src/reducers/account'
import { RootState } from 'src/reducers'
import { getDocumentClaims } from 'src/utils/filterDocuments'
import { ThunkDispatch } from 'src/store'

import I18n from 'src/locales/i18n'
import { filters } from 'src/lib/filterDecoratedClaims'

import { DocumentsCarousel } from '../components/documentsCarousel'
import { DocumentsList } from '../components/documentsList'
import { DocumentViewToggle } from '../components/documentViewToggle'
import strings from '../../../locales/strings'
import { Typography, Colors } from 'src/styles'
import { Wrapper } from '../../structure'

interface Props
  extends ReturnType<typeof mapDispatchToProps>,
    ReturnType<typeof mapStateToProps> {}

interface State {
  activeDocumentIndex: number
  showingValid: boolean
}

const styles = StyleSheet.create({
  emptyDocumentsText: {
    ...Typography.mainText,
    ...Typography.centeredText,
    color: Colors.greyLight,
    paddingHorizontal: '5%',
  },
})

export class DocumentsContainer extends React.Component<Props, State> {
  public ScrollViewRef: ScrollView | null = null

  public state = {
    activeDocumentIndex: 0,
    showingValid: true,
  }

  private scrollToTop(): void {
    if (this.ScrollViewRef) {
      this.ScrollViewRef.scrollTo({ x: 0, y: 0, animated: true })
    }
  }

  private onActiveIndexChange(index: number): void {
    this.setState({ activeDocumentIndex: index })
    this.scrollToTop()
  }

  private handleToggle = () => {
    this.setState((prevState: State) => ({
      showingValid: !prevState.showingValid,
      activeDocumentIndex: 0,
    }))
    this.scrollToTop()
  }

  public render(): JSX.Element {
    const { openDocumentDetails, decoratedCredentials } = this.props
    const documents = getDocumentClaims(decoratedCredentials['Other'])
    const docFilter = this.state.showingValid
      ? filters.filterByValid
      : filters.filterByExpired
    const displayedDocs = docFilter(documents)
    const isEmpty = displayedDocs.length === 0
    const otherIsEmpty = displayedDocs.length === documents.length

    if (isEmpty) {
      return (
        <Wrapper centered>
          <Text style={styles.emptyDocumentsText}>
            {I18n.t(strings.NO_DOCUMENTS_TO_SEE_HERE) + '...'}
          </Text>
        </Wrapper>
      )
    }

    return (
      <Wrapper>
        <Animated.View style={{ flex: 1 }}>
          {!otherIsEmpty && (
            <DocumentViewToggle
              showingValid={this.state.showingValid}
              onTouch={this.handleToggle}
            />
          )}
          <ScrollView
            // to scroll to top upon changing card
            ref={ref => (this.ScrollViewRef = ref)}
            contentContainerStyle={
              isEmpty ? { flex: 1 } : { paddingBottom: 110 }
            }
            scrollEnabled={!isEmpty}
          >
            {this.state.showingValid ? (
              <DocumentsCarousel
                documents={displayedDocs}
                activeIndex={this.state.activeDocumentIndex}
                onActiveIndexChange={this.onActiveIndexChange.bind(this)}
              />
            ) : (
              <DocumentsList
                documents={displayedDocs}
                onDocumentPress={openDocumentDetails}
              />
            )}
          </ScrollView>
        </Animated.View>
      </Wrapper>
    )
  }
}

const mapStateToProps = ({
  account: {
    claims: { decoratedCredentials },
  },
}: RootState) => ({
  decoratedCredentials,
})

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  openDocumentDetails: (doc: DecoratedClaims) =>
    dispatch(openDocumentDetails(doc)),
})

export const Documents = connect(
  mapStateToProps,
  mapDispatchToProps,
)(DocumentsContainer)
