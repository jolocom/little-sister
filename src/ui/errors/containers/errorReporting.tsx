import React, { useState } from 'react'
import { connect } from 'react-redux'
import { NavigationScreenProp, NavigationState } from 'react-navigation'
import { AppError, reportError } from '../../../lib/errors'
import { ThunkDispatch } from '../../../store'
import { routeList } from '../../../routeList'
import { navigationActions } from '../../../actions'
import { Emoji, EmojiSection } from '../components/emojiSection'
import { JolocomButton, Wrapper } from '../../structure'
import { ScrollView } from 'react-native'
import { ChooseIssueSection } from '../components/chooseIssueSection'
import { DescriptionSection } from '../components/descriptionSection'
import { ContactSection } from '../components/contactSection'
import I18n from '../../../locales/i18n'
import strings from '../../../locales/strings'
import { SectionWrapper } from '../components/sectionWrapper'
import { NavigationSection } from '../components/navigationSection'

interface PaymentNavigationParams {
  error?: AppError | Error
  previousScreen: routeList
}

interface Props extends ReturnType<typeof mapDispatchToProps> {
  navigation: NavigationScreenProp<NavigationState, PaymentNavigationParams>
}

export enum Inputs {
  None,
  Dropdown,
  Description,
  Contact,
}

const ErrorReportingContainer = (props: Props) => {
  const { navigateToScreen, navigation } = props

  const [pickedIssue, setIssue] = useState('')
  const [description, setDescription] = useState('')
  const [contact, setContact] = useState('')
  const [currentInput, setInput] = useState(Inputs.None)
  const [toggleState, setToggle] = useState(false)
  const [selectedEmoji, setEmoji] = useState(Emoji.Empty)

  // NOTE send previous screen as props, since not all
  // error reports will have error objects

  // const { previousScreen } = navigation.state.params
  // const isBackButton = previousScreen === routeList.Settings
  const isBackButton = false

  // NOTE error reports can show up without an error,
  // hence a navigation prop (navigateTo) would be better than in the error object
  const navigateBack = () => {
    const error = props.navigation?.state?.params?.error
    const previousScreen = props.navigation?.state.params?.previousScreen
    if (error instanceof AppError) {
      navigateToScreen(error.navigateTo)
    } else {
      navigateToScreen(previousScreen || routeList.AppInit)
    }
  }

  const onSubmitReport = () => {
    const error = props.navigation?.state?.params?.error

    const userReport = {
      userError: pickedIssue,
      userDescription: description,
      userContact: contact,
      sendPrivateData: toggleState,
    }

    if (navigation) {
      reportError({ ...userReport, error })
      navigateBack()
    }
  }

  return (
    <Wrapper secondaryDark>
      <NavigationSection
        onNavigation={navigateBack}
        isBackButton={isBackButton}
      />
      <ScrollView overScrollMode="never">
        <SectionWrapper
          title={I18n.t(strings.TELL_US_THE_PROBLEM)}
          style={{ marginTop: 14 }}>
          <ChooseIssueSection
            currentInput={currentInput}
            pickedIssue={pickedIssue}
            onIssuePick={setIssue}
            setInput={setInput}
          />
        </SectionWrapper>
        <SectionWrapper title={I18n.t(strings.CAN_YOU_BE_MORE_SPECIFIC)}>
          <DescriptionSection
            currentInput={currentInput}
            setInput={setInput}
            setDescription={setDescription}
            toggleState={toggleState}
            setToggle={setToggle}
            description={description}
          />
        </SectionWrapper>
        <SectionWrapper title={I18n.t(strings.WANT_TO_GET_IN_TOUCH)}>
          <ContactSection
            onContactInput={setContact}
            currentInput={currentInput}
            setInput={setInput}
            contactValue={contact}
          />
        </SectionWrapper>
        <SectionWrapper title={I18n.t(strings.RATE_THE_ISSUE)}>
          <EmojiSection selectedEmoji={selectedEmoji} setEmoji={setEmoji} />
        </SectionWrapper>
        <JolocomButton
          onPress={onSubmitReport}
          containerStyle={{
            marginTop: 45,
            marginBottom: 66,
            marginHorizontal: 20,
            height: 56,
          }}
          text={I18n.t(strings.SUBMIT_REPORT)}
        />
      </ScrollView>
    </Wrapper>
  )
}

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  navigateToScreen: (screen: routeList) =>
    dispatch(navigationActions.navigate({ routeName: screen })),
})

export const ErrorReporting = connect(
  null,
  mapDispatchToProps,
)(ErrorReportingContainer)
