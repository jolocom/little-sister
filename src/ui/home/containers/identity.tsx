import React from 'react'
import { IdentityComponent } from 'src/ui/home/components/identity'
import { connect } from 'react-redux'
import { RootState } from 'src/reducers/'

interface ConnectProps { }

interface Props extends ConnectProps {}

interface State {
  userName: string,
  phoneNumber: string,
  emailAddress: string,
  scanning: boolean
}

export class IdentityContainer extends React.Component<Props, State> {

  state = {
    userName: '',
    phoneNumber: '',
    emailAddress: '',
    scanning: false
  }

  componentDidMount() {
  }
  private onUserNameChange= (userName: string) : void => {
    this.setState({ userName })
  }

  private onPhoneNumberChange= (phoneNumber: string) : void => {
    this.setState({ phoneNumber })
  }

  private onEmailAddressChange= (emailAddress: string) : void => {
    this.setState({ emailAddress })
  }

  private onScannerStart = () : void => {
    this.setState({ scanning: true })
  }

  private onScannerCancel = () : void => {
    this.setState({ scanning: false })
  }

  // TODO Typings on E, event is not enough
  private onScannerSuccess = (e : Event) : void => {
    this.setState({ scanning: false })
  }

  render() {
    return (
    <IdentityComponent
      scanning={ this.state.scanning }
      userName= {this.state.userName}
      phoneNumber= { this.state.phoneNumber }
      emailAddress= { this.state.emailAddress }
      onUserNameChange= { this.onUserNameChange }
      onPhoneNumberChange= { this.onPhoneNumberChange }
      onEmailAddressChange= { this.onEmailAddressChange }
      onScannerStart={ this.onScannerStart }
      onScannerSuccess={ this.onScannerSuccess }
      onScannerCancel={ this.onScannerCancel }
     />
    )
  }
}

const mapStateToProps = (state: RootState) => {
  return {
  }
}

const mapDispatchToProps = (dispatch: (action: Function) => void) => {
  return {
  }
}

export const Identity = connect(mapStateToProps, mapDispatchToProps)(IdentityContainer)
