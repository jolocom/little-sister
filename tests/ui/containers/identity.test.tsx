import React from 'react'
import { shallow } from 'enzyme'
import { IdentityContainer } from 'src/ui/home/containers/identity'


describe('Identity container', () => {
  const COMMON_PROPS = {
    userName: '',
    phoneNumber: '',
    emailAddress: '',
    scanning: false
  }

  it('mounts correctly and matches the snapshot', () => {
    const rendered = shallow(<IdentityContainer {...COMMON_PROPS}/>)
    expect(rendered).toMatchSnapshot()
  })

  it('correctly changes scanning to true when qr code scanner is started', () => {
    const onScannerStart = jest.fn()

    const props = Object.assign({}, COMMON_PROPS {
      scanning: true
    })

    const rendered = shallow(<IdentityContainer {...props}/>)
    expect(rendered).toMatchSnapshot()
  })

  it('correctly changes scanning to false when qr code scanner is canceled', () => {
    const onScannerCancel = jest.fn()
    const props = Object.assign({}, COMMON_PROPS {
      scanning: false
    })

    const rendered = shallow(<IdentityContainer {...props}/>)
    expect(rendered).toMatchSnapshot()
  })
