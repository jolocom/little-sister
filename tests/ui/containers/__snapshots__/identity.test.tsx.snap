// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Identity container correctly changes scanning to false when qr code scanner is canceled 1`] = `
<IdentityComponent
  emailAddress=""
  onEmailAddressChange={[Function]}
  onPhoneNumberChange={[Function]}
  onScannerCancel={[Function]}
  onScannerStart={[Function]}
  onScannerSuccess={[Function]}
  onUserNameChange={[Function]}
  phoneNumber=""
  scanning={false}
  userName=""
/>
`;

exports[`Identity container correctly changes scanning to true when qr code scanner is started 1`] = `
<IdentityComponent
  emailAddress=""
  onEmailAddressChange={[Function]}
  onPhoneNumberChange={[Function]}
  onScannerCancel={[Function]}
  onScannerStart={[Function]}
  onScannerSuccess={[Function]}
  onUserNameChange={[Function]}
  phoneNumber=""
  scanning={false}
  userName=""
/>
`;

exports[`Identity container mounts correctly and matches the snapshot 1`] = `
<IdentityComponent
  emailAddress=""
  onEmailAddressChange={[Function]}
  onPhoneNumberChange={[Function]}
  onScannerCancel={[Function]}
  onScannerStart={[Function]}
  onScannerSuccess={[Function]}
  onUserNameChange={[Function]}
  phoneNumber=""
  scanning={false}
  userName=""
/>
`;
